src/blockchain/transactions.rs

```diff
@@ -14,155 +14,147 @@
-        };
-
-        let verification_result = input_signer.verify_input(
-            TxInRef::new(tx.as_ref(), self.input as usize),
-            input_tx.as_ref(),
-            &public_key,
-            input_signature_ref,
-        );
-
-        if verification_result.is_err() {
-            return Err(SignatureError::VerificationFailed.into());
+        // Make sure txid is equal to the identifier of the anchoring transaction proposal.
+        if proposal.id() != arg.txid {
+            return Err(Error::UnexpectedProposalTxId.into());
         }
 
-        let input_id = self.input_id();
+        // Check that input signature is correct.
+        let redeem_script = actual_config.redeem_script();
+        let input_signer = InputSigner::new(redeem_script.clone());
+        arg.verify_signature(&input_signer, &public_key, &proposal, &expected_inputs)?;
+
+        // All preconditions are correct and we can use this signature.
+        let input_id = TxInputId::new(proposal.id(), arg.input);
         let mut input_signatures = schema.input_signatures(&input_id, &redeem_script);
-        if input_signatures.len() != redeem_script_content.quorum {
-            // Adds signature to schema.
-            input_signatures.insert(self.validator, self.input_signature.clone().into());
+        let quorum = redeem_script.content().quorum;
+        let mut input_signature_len = input_signatures.len();
+        // Check that we have not reached the quorum yet, otherwise we should not do anything.
+        if input_signature_len < quorum {
+            // Add signature to schema.
+            input_signatures.insert(anchoring_node_id, arg.input_signature.clone().into());
             schema
                 .transaction_signatures()
                 .put(&input_id, input_signatures);
-            // Tries to finalize transaction.
-            let mut tx: btc::Transaction = tx.clone();
+            input_signature_len += 1;
+        } else {
+            return Ok(());
+        }
+
+        // If we have enough signatures for specific input we have to check that we also have
+        // sufficient signatures to finalize proposal transaction.
+        if input_signature_len == quorum {
+            let mut finalized_tx: btc::Transaction = proposal.clone();
+            // Make sure we reach a quorum for each input.
             for index in 0..expected_inputs.len() {
-                let input_id = TxInputId::new(self.transaction.id(), index as u32);
+                let input_id = TxInputId::new(proposal.id(), index as u32);
                 let input_signatures = schema.input_signatures(&input_id, &redeem_script);
-
-                if input_signatures.len() != redeem_script_content.quorum {
+                // We have not enough signatures for this input, so we can not finalize this
+                // proposal at the moment.
+                if input_signatures.len() != quorum {
```

Comment:
`input_signatures` here are different from `input_signatures` above.
Especially in the diff it looks really weird: 
```rust
if input_signature_len == quorum {
 ...
 if input_signatures.len() != quorum {
 ...
```

I find it really confusing.